### DISCUSSION TITLE
Duplicates not resolving (initial upload via CLI)

### DISCUSSION THREAD

**Discussion by wutr** [immich-app/immich#2109] [Q&A]
After finally setting up immich I imported all 8000+ photos and videos from my android phone backup using the CLI. I then installed the immich android app a couple of days later. I've selected one album to be automatically backed up: camera. It contains 1401 items.

The thing is, apart from the photos taken in-between importing via the CLI and installing the android app, all other items show twice. One with a cloud icon (no tick), and one with a crossed out cloud icon. The files are duplicates, not reduced resolution versions or anything like that.

I was under the impression an automatic duplicate check is carried out but that seems to have not been the case.

Am I missing something here?

**Comment by etnoy** [comment]
Deduplication is currently being worked on, but in the end what decides is the file checksum. If the checksum already exists the file won't be uploaded. Any changes like changed metadata will allow a duplicate upload.

Could you compare the sha1 checksums before they are uploaded and post the result?

**Reply by wutr** [reply]
Thanks for the quick reply. I've compared the sha1 sums for one file that shows up twice in the immich android app. The android original, the file in my backup, and the file as it is after being imported in immich are all identical:

```
7184f793d0e9351ce6e41ac436e06766c2900191  android/IMG_20230322_161643.jpg
7184f793d0e9351ce6e41ac436e06766c2900191  backup/IMG_20230322_161643.jpg
7184f793d0e9351ce6e41ac436e06766c2900191  immich/IMG_20230322_161643.jpg
```

**Comment by alextran1502** [comment]
There are some sync issues on the mobile app, but they are mostly cosmetic. Can you confirm that the duplication also shows on the web?

**Reply by wutr** [reply]
Thanks for the quick reply. The web app does not have the duplicates.

Does that mean the android app has not attempted syncing the files that show as duplicates, and only synced/uploaded the files created after the CLI import was done?

**Reply by alextran1502** [reply]
Are you on the latest mobile app version v1.52? We solved a few local synchronization issues on the mobile app in the yesterday release

**Reply by wutr** [reply]
Not yet, I only installed the android app about 3 hours before I started this thread and the latest had not yet been released (or at least I wasn't aware). I noticed the notification when I did the SHA1 sums earlier but havent yet checked the release notes. I'll get back to you once I've updated to the latest server and app.

**Reply by nealhead** [reply]
I have this same issue. I just installed today and did CLI backup to load my initial images and now my app shows duplicates with the local image and the remote image that are the same. I did notice the following error in the logs, so it appears that it knows it's the same picture. App version 1.52.0 build 75. Server 1.52.1.

2023-03-29 22:57:09.248 EDT [13735] ERROR:  duplicate key value violates unique constraint "UQ_userid_checksum"
2023-03-29 22:57:09.248 EDT [13735] DETAIL:  Key ("ownerId", checksum)=(98844871-00e2-4ed0-a9cf-e2bf7b35183e, \x7f4f6331c50d6dd87c95f48ceedd9b7318dc8ee0) already exists.
2023-03-29 22:57:09.248 EDT [13735] STATEMENT:  INSERT INTO "assets"("id", "deviceAssetId", "ownerId", "deviceId", "type", "originalPath", "resizePath", "webpPath", "encodedVideoPath", "createdAt", "updatedAt", "fileCreatedAt", "fileModifiedAt", "isFavorite", "mimeType", "checksum", "duration", "isVisible", "livePhotoVideoId") VALUES (DEFAULT, $1, $2, $3, $4, $5, $6, $7, $8, DEFAULT, DEFAULT, $9, $10, $11, $12, $13, $14, $15, DEFAULT) RETURNING "id", "webpPath", "encodedVideoPath", "createdAt", "updatedAt", "isFavorite", "isVisible"

**Reply by wutr** [reply]
The issue persists with the latest Android app and server versions. I haven't tried recreating the library yet (ie. start from scratch).

I can't upload all photos from my phone as it doesn't have enough storage to accommodate them all, hence my interest in immich.

**Reply by alextran1502** [reply]
Can you help confirm that the duplication is only shown on the mobile app, is the duplication only shown on the web? Does it help resolve the issue if you logout of the mobile app and log back in?

**Reply by nealhead** [reply]
> Can you help confirm that the duplication is only shown on the mobile app, is the duplication only shown on the web? Does it help resolve the issue if you logout of the mobile app and log back in?

For me, the duplicates are only local. One picture has the cloud, and one has one with the cross through it. If I take a new picture it backs up as expected.

**Reply by wutr** [reply]
Likewise. The pictures don't appear twice (duplicated) on the web.

Logging in- and out of the mobile app doesn't change anything.

I have only one album selected to be backed up in the app: "camera". It contains 1403 items. Most of those were imported to immich via the CLI, and about 150 were imported/backed up via the app. The app says there are 157 unique assets (likely the number of pictures/videos taken since the initial CLI upload).

**Reply by alextran1502** [reply]
Have you uploaded those photos yet? The server/app won't know about duplication if the files haven't been uploaded to the server yet.

**Reply by wutr** [reply]
The files created after the CLI import are automatically uploaded by the immich android app and show in the web UI. They have the cloud-with-tick icon.

The files that are on the phone in the "camera" album as well as in the CLI upload have not been uploaded separately via the immich app afterwards if that's what you mean. These files are shown twice in the android app: one with a cloud, and one with a crossed-out cloud.

**Reply by alextran1502** [reply]
The files with the crossed-out cloud mean they are on your mobile device and have not been uploaded yet, so they could be shown next to the already uploaded file and give you the indication they are duplicated. It should resolve the display of the files once those are uploaded.

**Reply by nealhead** [reply]
> The files with the crossed-out cloud mean they are on your mobile device and have not been uploaded yet, so they could be shown next to the already uploaded file and give you the indication they are duplicated. It should resolve the display of the files once those are uploaded.

But what if those photos were uploaded already via CLI? I see the error in the log showing they are a duplicate but it still shows it twice instead of showing that they are already uploaded.

**Reply by wutr** [reply]
> > The files with the crossed-out cloud mean they are on your mobile device and have not been uploaded yet, so they could be shown next to the already uploaded file and give you the indication they are duplicated. It should resolve the display of the files once those are uploaded.
> 
> But what if those photos were uploaded already via CLI? I see the error in the log showing they are a duplicate but it still shows it twice instead of showing that they are already uploaded.

That, and also the fact that the immich android app makes no attempt to upload the files nor is there an option to manually do it. The backup information shows there are no assets remaining to be uploaded.

**Reply by alextran1502** [reply]
The way it should work for this scenario is that it would upload the file to the server, receive the duplication status, and then put it into the local database marked as duplicated and not show it on the timeline. 

@fyfrey Could it be a bug with this scenario for the sync mechanism?

**Reply by fyfrey** [reply]
It's a mix of everything: assets from the mobile app were uploaded from the CLI, resulting in a different device ID. Thus, the sync cannot match these assets with the locally existing assets.
The app also can't upload because the server says these are already uploaded (via SHA1 hash match).
We could mark assets that are duplicates as uploaded in the app. However, doing so leads to issues once these images with matching SHA1 are deleted on the server. The app would still say the assets are saved on the server which is not true. So.. this is even worse.
The only proper solution is to also hash on the mobile app, which we are already working on.

**Reply by wutr** [reply]
Thank you, so this is intended behaviour for now then?

**Reply by alextran1502** [reply]
I would say yes, until the client-side hashing is implemented

**Reply by alextran1502** [reply]
The workaround is to choose a different album with one or two assets so that the local assets don't get rendered on the main timeline.

**Reply by wutr** [reply]
> The workaround is to choose a different album with one or two assets so that the local assets don't get rendered on the main timeline.

Won't that prevent new files from being synced to the server if they're not part of that dummy album?

**Reply by fyfrey** [reply]
yes. that would prevent new files from being backed up.
you could move all assets already backed up via CLI to a new device album (and not select this in Immich). Then your newly taken photos still go to the Camera album (and you select only that to be backed up)

**Reply by wutr** [reply]
I think I'll nuke-and-pave my immich setup, and do the following:

1. Reset the immich android app and let it upload all desired items (the camera album)
2. Once finished, use jdupes to identify which pictures from my backup were not present on my phone and create a symlink to each file in a temporary directory.
3. Use immich CLI to upload all the symlinked items

I'll have to double check the volume mapping allows for my symlink strategy, I'll probably run into some issues.
